src/lang/translation_provider_FILE.hpp

```diff
@@ -0,0 +1,96 @@
+#pragma once
+
+#include "translation_provider.hpp"
+#include <stdio.h>
+#include <string.h>
+#include "gettext_string_hash.hpp"
+
+//TODO fix including from string.h
+extern "C" size_t strlcpy(char *dst, const char *src, size_t dsize);
+
+class FILEtranslationProvider : public ITranslationProvider {
+
+    uint16_t findMessage(const char *key, FILE *file) const {
+
+        uint32_t index = m_HashTable.GetIndexOfKey(key);
+
+        //nothing is on index 0
+        if (index == 0) {
+            return false;
+        }
+
+        //get to position of translated string
+        fseek(file, m_TransTableOff + (8 * index) + 4, SEEK_SET);
+
+        uint32_t posOfTrans = 0;
+        fread(&posOfTrans, 4, 1, file);
+        fseek(file, posOfTrans, SEEK_SET);
+        return posOfTrans;
+    }
+
+    char m_Path[16];
+    FILE *m_File = nullptr;
+    gettext_hash_table m_HashTable;
+    uint32_t m_TransTableOff;
+
+public:
+    FILEtranslationProvider(const char *path) {
+#ifdef TRANSLATIONS_UNITTEST
+        strncpy(m_Path, path, sizeof(m_Path));
+#else
+        strlcpy(m_Path, path, sizeof(m_Path));
+#endif
+    }
+
+    ~FILEtranslationProvider() override = default;
+    string_view_utf8 GetText(const char *key) const override {
+
+        //check if file is valid, if not try to open it again
+        if (!m_File) {
+            if (!(const_cast<FILEtranslationProvider *>(this)->OpenFile())) {
```

Comment:
If the provider needs to call non-const methods here we shall think about improving the interface.
The `const_cast` usually means there is something seriously wrong with the code